pkg/arvo/app/metadata-hook.hoon

```diff
@@ -0,0 +1,200 @@
+      [~ state]
+    :_  state(synced (~(del by synced) path.act))
+    %-  zing
+    :~  (unsubscribe [%group path.act] u.ship)
+        ?.  &(=(u.ship our.bowl) (team:title our.bowl src.bowl))  ~
+        [%give %kick ~[[%group path.act]] ~]~
+    ==
+  ==
+  ::
+  ++  unsubscribe
+    |=  [=path =ship]
+    ^-  (list card)
+    ?:  =(ship our.bowl)
+      [%pass path %agent [our.bowl %metadata-store] %leave ~]~
+    [%pass path %agent [ship %metadata-hook] %leave ~]~
+  --
+::
+++  watch-group
+  |=  =path
+  ^-  (list card)
+  |^
+  ?>  (~(has by synced) path)
+  ?>  (is-permitted [(scot %p src.bowl) path])
+  %+  turn  ~(tap by (metadata-scry path))
+  |=  [[=group-path =resource] =metadata]
+  ^-  card
+  [%give %fact ~ %metadata-update !>([%add group-path resource metadata])]
+  ::
+  ++  is-permitted
+    |=  pax=^path
+    ^-  ?
+    =.  pax
+      ;:(weld /=permission-store/(scot %da now.bowl)/permitted pax /noun)
+    .^(? %gx pax)
+  ::
+  ++  metadata-scry
+    |=  pax=^path
+    ^-  associations
+    =.  pax  ;:(weld /=metadata-store/(scot %da now.bowl)/group pax /noun)
+    (need .^((unit associations) %gx pax))
+  --
+::
+++  fact-metadata-update
+  |=  [wir=wire fact=metadata-update]
+  ^-  (quip card _state)
+  |^
+  [?:((team:title our.bowl src.bowl) handle-local handle-foreign) state]
+  ::
+  ++  handle-local
+    ?+  -.fact  ~
+        %add
+      ?.  (~(has by synced) group-path.fact)  ~
+      (give group-path.fact fact)
+    ::
+        %update-metadata
+      ?.  (~(has by synced) group-path.fact)  ~
+      (give group-path.fact fact)
+    ::
+        %remove
+      ?.  (~(has by synced) group-path.fact)  ~
+      (give group-path.fact fact)
+    ==
+  ::
+  ++  handle-foreign
+    ?+  -.fact  ~
+        %add
+      ?.  =(src.bowl (~(got by synced) group-path.fact))  ~
+      (poke fact)
+    ::
+        %update-metadata
+      ?.  =(src.bowl (~(got by synced) group-path.fact))  ~
+      (poke [%add +.fact])
+    ::
+        %remove
+      ?.  =(src.bowl (~(got by synced) group-path.fact))  ~
+      (poke fact)
+    ==
+  ::
+  ++  give
+    |=  [=path upd=metadata-update]
+    ^-  (list card)
+    [%give %fact ~[[%group path]] %metadata-update !>(upd)]~
+  ::
+  ++  poke
+    |=  act=metadata-action
+    ^-  (list card)
+    [%pass / %agent [our.bowl %metadata-store] %poke %metadata-action !>(act)]~
+  --
+::
+++  kick
+  |=  wir=wire
+  ^-  (quip card _state)
+  :_  state
+  ?+  wir  !!
+      [%updates ~]
+    [%pass /updates %agent [our.bowl %metadata-store] %watch /updates]~
+  ::
+      [%group @ *]
+    ?.  (~(has by synced) t.wir)  ~
+    =/  =ship  (~(got by synced) t.wir)
+    ?:  =(ship our.bowl)
+      [%pass wir %agent [our.bowl %metadata-store] %watch wir]~
+    [%pass wir %agent [ship %metadata-hook] %watch wir]~
+  ==
+::
+++  watch-ack
+  |=  [wir=wire saw=(unit tang)]
+  ^-  (quip card _state)
+  ?>  ?=(^ wir)
+  [~ ?~(saw state state(synced (~(del by synced) t.wir)))]
```

Comment:
Don't we want to do the same `%kick`ing of incoming subscriptions we do in the `%remove` action case?

Reply:
No, because we aren't going to receive a "not-permitted" case if we are the host of the resource, and we would only kick subscriptions if we were the host of the resource.

Reply:
Okay, I think I get that.

It looks like this `+watch-ack` only works for the group subscriptions, though, because of the assertion on the wire. We use `/updates` as our wire for watching the metadata-store, but use this for handling watch-acks always, so I'm a bit confused on how this doesn't insta-crash.

Reply:
The wir would be a cell? Even in the /updates case? [%updates ~]

Reply:
Man, I shouldn't be doing review comments early in the morning.

But then, `(~(del by synced) ~)` doesn't make a lot of sense. We should definitely check for `?=([%group ^] wir)` instead, and only do the deletion logic in that case, so it's clear that that subscription flow is what this logic belongs to.